### ISSUE TITLE
MNIST example Error

### ISSUE THREAD

**Issue by myhussien** [vlfeat/matconvnet#85]
When I run the MNIST example shipped by the toolbox using the default parameters, it works just fine. However, when I change the first conv. layer weights defined in (cnn_mnist) from 5x5x1x20 to 9x9x1x20 or any number larger than 5 for the first two dimensions, this error shows up:

FILTERS are larger than the DATA (including padding)

I checked the size of the input in a batch and I found it 28x28x1x100. Since the width and the height of the filter are smaller than the input, there shouldn't be such error. Can you please check this or correct me if I'm wrong?

Thank you!

-MYH

**Comment by drohit92** [comment]
Hi , the problem is not in the 1st layer but in the filters of the 5th layer. Since you changed the filters of the 1st layer to 9x9x1x20, the output now coming out of the 4th layer is 3x3x50x100.However the filters of the 5th layer i.e. 4x4x50x500 remained unchanged. As the filter is greater than the data in this case, hence the error.

Simply changing the filters of the 5th layer to 3x3x50x500 will solve your problem.

In the future, I suggest you make use of MATLAB's debugging tool. It's simple to use and you can easily put break points and pin point the exact problem.

Hope it helps.

**Comment by myhussien** [comment]
Much appreciated! Would you mind explaining  the size computation from the input to the 5th layer?   I think I'm doing things wrong.

**Comment by drohit92** [comment]
Hi,

If we go by your architecture, the entire network would look something like this:

Notations used below: 

1. conv9-20 => Filter is of size 9x9 and there are 20 output activation maps
2. pool 2 => 2x2 pooling
3. S=1 => stride of 1
4. P=0 => padding of 0

Formula for calculating output size for a conv layer: (W+2P-F)/S+1 where W is input size
For a Relu Layer input and output are of same sizes
For a pooling layer the input size get divided by pooling size e.g. 2 in case of 2*2

Input Data: 28x28x1x100

Input: 28x28x1x100    -> Layer 1: conv9-20 S=1 P=0    -> Output: 20x20x20x100 
Input: 20x20x20x100  -> Layer 2: Pool 2 S=2 P=0         -> Output: 10x10x20x100
Input: 10x10x20x100  -> Layer 3: conv5-50 S=1 P=0    -> Output:  6x6x50x100 
Input: 6x6x50x100      -> Layer 4: Pool 2 S=2 P=0         -> Output: 3x3x50x100
Input: 3x3x50x100      -> Layer 5: conv3-500 S=1 P=0  -> Output:  1x1x500x100 
Input: 1x1x500x100    -> Layer 6: ReLU                        -> Output:  1x1x500x100 
Input: 1x1x500x100    -> Layer 7: conv1-10 S=1 P=0    -> Output:  1x1x10x100
Input: 1x1x10x100      -> Layer 8: Softmax Loss            -> Output:  Loss

Note: 100 here is the batch size. So the final output of layer 7 has 10 values. One for each class (0-9) and in total 100 such values for each training image in batch.

I hope I didn't confuse you.

**Comment by myhussien** [comment]
Can explain the output size equation by doing the 1st and the second layers as examples?  F  is not defined as well.

**Comment by drohit92** [comment]
Layer1: Input is 28x28x1x100  Filter Details: conv9-20 S=1 P=0

W(Input height of image)=28, F(size of filter)=9, Stride=1, Padding=0

For convolution output is: (W+2P-F)/S+1 => (28+2*0-9)/1+1=20

Similarliy for width, output size is  (28+2*0-9)/1+1=20

Hence the total output size is 20x20

Now since the weights were 9x9x1x20 => conv will be between 28x28x1 and 9x9x1. This will give 20x20x1 output (explained above). Also there will be 20 output activation maps.

Hence final output of layer 1 for 1 image is: 20x20x20
since batch size is 100. So final output now is 20x20x20x100 

Next is pooling. size is just half of layer 1 output. So layer 2 output is 10x10x20x100. 

Now layer 3: weights are 5x5x20x50. Conv will be between 10x10x20 and 5x5x20. You can calculate and confirm that output size will be 6x6x20. Since 50 output activation maps are required, so final output of layer3 for 1 image is  6x6x50, and similarly for 100 batch size output will be 6x6x50x100

**Comment by myhussien** [comment]
I can't thank you enough! I have one final question. If the images were of size 28x28x3x100 what is the difference between applying a filter of 9x9x1x20, 9x9x20x1, and 9x9x20x20? My main issue is with last 2 dimensions. Also do we need to rescale the input to the 0-1 range of a Softmax layer is applied at the end, or does the CNN take care of this?

**Comment by drohit92** [comment]
The filter dimesions are of the format: [filter width] x [filter height] x [number of input channels] x [number of output activation maps required]

So if the data of size 28x28x1x100 is given to layer1 and I have to decide the weight dimensions, then I will first decide what size filters I want. Let's say 9x9, then my weight dimesions would be like 9x9x__x__

Now since my input image has only 1 channel (black and white MNIST images) ,(3 in case of RGB images), hence the layer 1 has 3 input channels. SO my weight matrix becomes now 9x9x1x__

Now I decide how mnay output activation maps I require. Can be any number. In this case 20. So my weight matrix becomes 9x9x1x20

Input: 28x28x1x100,   W: 9x9x1x20,  Output: 20x20x20x100  [The first 20x20 is because of the convolution, and the last 20 because of the no.of output activation maps you decided upon in previous step]

The input should always be normalised. I guess in the MNIST the mean is subtracted something like this:

dataMean = mean(data(:,:,:,set == 1), 4);
data = bsxfun(@minus, data, dataMean) ;

I do not know if it should be rescaled to 0-1 or not.

**Comment by myhussien** [comment]
Thank you! I understand this part. My question is, what would you do if the filter for example was 9x9x10x20 instead? now it is 20 activation maps, would you still have 100 in the 4th output dimension? also now you have 10 filter channels. How do you apply them to the input? in 3D fashion?

**Comment by drohit92** [comment]
The 100 is the batch size  and comes up only in the output. Let me try and explain again.

The input is 28x28x1x100 where image size is 28x28 of just one channel and 100 such images.

Ok now I have a filter say 9x9x10x20. I cannot convolve this filter with the image because the matrix dimesions don't match.

I can convolve one image 28x28 with 9x9 filter resulting in 20x20, but the 10 is incompatible with input dimesions of 1.

Remember to multiply 2 matrices of size axb and cxd, b should be equal to c

You can  try reading this tutorial to understand better about cnn weights etc http://cs231n.github.io/convolutional-networks/

**Comment by myhussien** [comment]
The link is great! Your help is greatly appreciated. I found what I was looking for in the link, mainly the following sentence " The connections are local in space (along width and height), but always full along the entire depth of the input volume". I didn't know that the depth of the filter should (always) equal to the input depth, and thought it can be applied as in medical 3D filtering- which is not the case. I enjoyed the discussion! 

Regards!
